### PR TITLE
Dependencies: use pip-tools for all our files 

### LINKED ISSUE #9273: pyup or dependabot together with pip-tools
pyup does not understand pip-tools (pip-compile), so I was taking a look at dependabot and it seems it does: https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/about-dependabot-version-updates#supported-repositories-and-ecosystems

We should probably migrate to it and use pip-tools for all our dependencies.

### PR DESCRIPTION
**Pull request by humitos**
Continuation of https://github.com/readthedocs/readthedocs.org/pull/9479 to use pip-tools for all our requirements files and the
automation of the GitHub Action.

----
:books: Documentation preview :books:: https://docs--9480.org.readthedocs.build/en/9480/

### CODE REVIEW

**Comment by humitos** [comment]
🎉 , everything is working fine!

It seems there is an issue with the migrations check, but definitely not related to this PR, unless the change from Django 3.2.13 to 3.2.15 affected it?

File: requirements/deploy.in [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Requirements for our production systems
+
+-r pip.txt
+
+psycopg2
```
Comment (benjaoming) [inline requirements/deploy.in:5]: This one used to be pinned `psycopg2==2.9.3`

File: requirements/deploy.in [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+gunicorn
```
Comment (benjaoming) [inline requirements/deploy.in:6]: This one used to be pinned `gunicorn==20.1.0`

File: requirements/deploy.in [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+django-redis-cache
+
+# For resizing images
+pillow
```
Comment (benjaoming) [inline requirements/deploy.in:10]: Used to be pinned as `pillow==9.1.0`, this one might be less risky to pin loosely?

Comment (benjaoming) [reply requirements/deploy.in:10]: (the update results in Pillow being bumped to 9.2.0)

File: requirements/pip.in [hunk 4]
```diff
@@ -0,0 +1,110 @@
+# Base packages
+pip
+virtualenv
+
+django~=3.2.0
```
Comment (benjaoming) [inline requirements/pip.in:5]: Used to be `<3.3` but I guess this is the same.

File: requirements/pip.in [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+django-extensions
+django-polymorphic
+django-autoslug
+django-simple-history
+
+djangorestframework
+
+# For intersphinx during builds
+Sphinx
```
Comment (benjaoming) [inline requirements/pip.in:14]: This used to be `Sphinx==4.5.0`

File: requirements/pip.in [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+# Filtering for the REST API
+django-filter
```
Comment (benjaoming) [inline requirements/pip.in:17]: This will update django-filter to `22.1` (used to be `21.1`)

**Review by benjaoming — COMMENTED** [review]
I didn't find a good systematic way to go through this, but I did find a lot of implicit updates.

I suppose that these dependencies didn't have a previous intention to be pinned to anything special since there were no comments next to them, but I just wanted to make sure that it's understood - welcome to click Resolve on any of these comments :+1:

File: requirements/pip.in [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+django-elasticsearch-dsl<8.0
+selectolax
+
+# embed
+pyquery
+
+# NOTE: this dep can be removed in python 3.7 in favor of ``date.fromisoformat``
+python-dateutil
+
+orjson
```
Comment (benjaoming) [inline requirements/pip.in:59]: This used to be `3.6.8` but will upgrade to `3.7.11`

**Comment by humitos** [comment]
> I suppose that these dependencies didn't have a previous intention to be pinned to anything special since there were no comments next to them, but I just wanted to make sure that it's understood

Yes. We are pinning _only the dependencies that we know can't be updated_ because the newer version breaks something, removes a feature we use or is incompatible with another dependency we rely on.

Before using pip-tools we were pinning _all the requirements_ in the `requirements.txt` file and control the upgrade of the version using pyup. However, we wanted to migrate to pip-tools which uses two different files: the `.in` file where the requirements do not have to be pinned (unless we have to force them as mentioned before) and the `.txt` file which pins all our dependencies at this moment in time and also pin its transitive dependencies.

This PR in particular contains upgrades on some versions because pyup has been broken for months and it was not running and updating the dependencies automatically. So, those upgrades are expected to be here as well. If there is an upgrade _we don't want to upgrade here_ we have to pin it, but I didn't find any case of these.

File: requirements/pip.in (hunk 4 above)
Comment (humitos) [reply requirements/pip.in:5]: Yes. It's the same.

File: requirements/pip.in [hunk 8]
```diff
@@ -0,0 +1,110 @@
+# Base packages
+pip
+virtualenv
+
+django~=3.2.0
+django-extensions
+django-polymorphic
+django-autoslug
+django-simple-history
```
Comment (benjaoming) [inline requirements/pip.in:9]: This used to be pinned at 3.0.0 and is now 3.1.1

3.1.0 has a lot of changes: https://django-simple-history.readthedocs.io/en/latest/#section-2

**Comment by benjaoming** [comment]
@humitos it's very understandable why this PR is needed and the approach chosen, love it :+1: I just had to dig around a little bit because there are actual upgrades in the diff that aren't obvious. If we could find a systematic approach to list them out, that'd perhaps be best?

File: requirements/pip.in (hunk 8 above)
Comment (humitos) [reply requirements/pip.in:9]: We can kept it pinned in this PR to avoid this problem and solve it in a different PR 👍🏼

**Comment by humitos** [comment]
> I just had to dig around a little bit because there are actual upgrades in the diff that aren't obvious. If we could find a systematic approach to list them, that'd perhaps be best?

It's hard to review because it's the first PR after some months without upgrading dependencies and also changing the tool used. However, in the future these PR's will be pretty small since they will be create once per week only when there is any change. So, I think it will be easier to review than this one.

pyup listed the links to the changelogs and project homepages. That was a neat feature that helped us to know what were the changes and if it was safe to merge. Unfortunately, pyup does not work as we want :/

**Comment by humitos** [comment]
I'm sorry for the force-push. I had some conflicts that I was sure how to solve them. I'm going to merge this PR into its base and solve all the problems there before merging it.

**Pull request merged**